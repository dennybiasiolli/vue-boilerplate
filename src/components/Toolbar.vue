<template>
  <v-toolbar
    fixed
    app
  >
    <v-toolbar-side-icon @click.stop="handleToggle" />
    <v-toolbar-title>
      Sample App!
    </v-toolbar-title>
    <v-spacer />
    <a
      v-if="isLogged"
      @click="handleLogout"
    >
      Logout
      <v-icon
        title="Logout"
      >
        exit_to_app
      </v-icon>
    </a>
  </v-toolbar>
</template>

<script>
import { createNamespacedHelpers } from 'vuex'
const { mapGetters, mapActions } = createNamespacedHelpers('auth')
const { mapState, mapMutations } = createNamespacedHelpers('a')

export default {
  name: 'Toolbar',
  computed: Object.assign({}, mapState(['drawer']), mapGetters(['isLogged'])),
  methods: Object.assign(
    {
      handleToggle() {
        this.toggleDrawer()
      },
      handleLogout() {
        return this.logoutAsync().then(() => {
          this.$router.push('/login')
        })
      }
    },
    mapMutations(['toggleDrawer']),
    mapActions(['logoutAsync'])
  )
}
</script>
