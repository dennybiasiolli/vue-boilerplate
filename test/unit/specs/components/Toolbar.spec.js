import Vuex from 'vuex'
import { shallow, createLocalVue } from '@vue/test-utils'
import Toolbar from '@/components/Toolbar'

const localVue = createLocalVue()
localVue.use(Vuex)

const state = {
  drawer: null
}

let getters, mutations, actions

// This function creates a new Vuex store
// instance for every new test case.
const createStore = () => {
  getters = {
    isLogged: jest.fn().mockReturnValue(true)
  }
  mutations = {
    toggleDrawer: jest.fn()
  }
  actions = {
    logoutAsync: jest.fn(() => Promise.resolve())
  }
  return new Vuex.Store({
    modules: {
      a: {
        namespaced: true,
        state,
        mutations
      },
      auth: {
        namespaced: true,
        actions,
        getters
      }
    }
  })
}

describe('Toolbar', () => {
  let component, vm, element

  beforeEach(() => {
    component = shallow(Toolbar, {
      localVue,
      store: createStore(),
      mocks: {
        $router: [],
        $route: {
          query: {
            redirect: undefined
          }
        }
      }
    })
    vm = component.vm
    element = component.element
  })

  test('should match snapshot', () => {
    expect(element).toMatchSnapshot()
  })

  // Evaluate the results of functions in
  // the raw component options
  test('sets the correct default data', () => {
    expect(typeof Toolbar.computed.drawer).toBe('function')
    expect(typeof Toolbar.methods.handleToggle).toBe('function')
    expect(typeof Toolbar.methods.toggleDrawer).toBe('function')
    expect(typeof Toolbar.methods.handleLogout).toBe('function')
    expect(typeof Toolbar.methods.logoutAsync).toBe('function')
  })

  describe('handleToggle', () => {
    test('should have been called when clicking on drawer button', () => {
      jest.spyOn(vm, 'handleToggle')
      jest.spyOn(vm, 'toggleDrawer')
      component.find('button').trigger('click')
      expect(vm.handleToggle).toHaveBeenCalled()
      expect(vm.toggleDrawer).toHaveBeenCalledWith()
    })
  })

  describe('logoutAsync', () => {
    test('should have been called when clicking on logout', () => {
      jest.spyOn(vm, 'handleLogout')
      jest.spyOn(vm, 'logoutAsync')
      component.find('a').trigger('click')
      // FIXME: unexplicable error in the next line
      // expect(vm.handleLogout).toHaveBeenCalled()
      expect(vm.logoutAsync).toHaveBeenCalledWith()
      return vm.logoutAsync().then(() => {
        expect(vm.$router).toEqual(['/login'])
      })
    })
  })
})
