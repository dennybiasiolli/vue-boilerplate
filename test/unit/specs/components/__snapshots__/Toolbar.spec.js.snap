// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar should match snapshot 1`] = `
<nav
  class="v-toolbar v-toolbar--fixed theme--light"
  style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);"
>
  <div
    class="v-toolbar__content"
    style="height: 64px;"
  >
    <button
      class="v-toolbar__side-icon v-btn v-btn--icon theme--light"
      type="button"
    >
      <div
        class="v-btn__content"
      >
        <i
          aria-hidden="true"
          class="v-icon material-icons theme--light"
        >
          menu
        </i>
      </div>
    </button>
     
    <div
      class="v-toolbar__title"
    >
      
    Sample App!
  
    </div>
     
    <div
      class="spacer"
    />
     
    <a>
      
    Logout
    
      <i
        aria-hidden="true"
        class="v-icon material-icons theme--light"
        title="Logout"
      >
        
      exit_to_app
    
      </i>
    </a>
  </div>
</nav>
`;
